underfs/cephfs/src/main/java/alluxio/underfs/cephfs/CephInputStream.java

```diff
@@ -57,67 +73,97 @@ public int available() throws IOException {
     }
 
-    long oldPos = mPos;
-    mPos = mMount.lseek(mFileHandle, targetPos, CephMount.SEEK_SET);
-    if (mPos < 0) {
-      int ret = (int) mPos;
-      mPos = oldPos;
-      throw new IOException("ceph.lseek: ret = " + ret);
+    long oldPos = mCephPos;
+    mCephPos = mMount.lseek(mFileHandle, targetPos, CephMount.SEEK_SET);
+    if (mCephPos < 0) {
+      int ret = (int) mCephPos;
+      mCephPos = oldPos;
+      throw new IOException("ceph.lseek: failed ret = " + ret);
     }
+    mBufValid = 0;
+    mBufPos = 0;
   }
 
   @Override
-  public long skip(long n) throws IOException {
+  public synchronized long skip(long n) throws IOException {
     if (n <= 0) {
       return 0;
     }
-    long targetPos = mPos + n;
+    long targetPos = getPos() + n;
     seek(targetPos);
     return n;
   }
 
   @Override
-  public int read() throws IOException {
-    if (-1 == read(mBuf, 0, 1)) {
+  public synchronized int read() throws IOException {
+    if (getPos() >= mFileLength) {
+      return -1;
+    }
+    if (-1 == read(mOneByteBuf, 0, 1)) {
       return -1;
     }
-    if (mBuf[0] < 0) {
-      return 256 + (int) mBuf[0];
+    if (mOneByteBuf[0] < 0) {
+      return 256 + (int) mOneByteBuf[0];
     } else {
-      return mBuf[0];
+      return mOneByteBuf[0];
     }
   }
 
   @Override
-  public int read(byte[] b, int off, int len) throws IOException {
-    if (b == null) {
+  public synchronized int read(byte[] buf, int off, int len) throws IOException {
+    if (mClosed.get()) {
+      throw new IOException(
+        "CephInputStream.read: cannot read " + len + "bytes from fd "
+        + mFileHandle + ": stream closed");
+    }
+    if (buf == null) {
       throw new NullPointerException();
-    } else if (off < 0 || len < 0 || len > b.length - off) {
+    } else if (off < 0 || len < 0 || len > buf.length - off) {
       throw new IndexOutOfBoundsException();
     } else if (len == 0) {
       return 0;
     }
 
     // ensure we're not past the end of the file
-    if (mPos >= mFileLength) {
+    if (getPos() >= mFileLength) {
       return -1;
     }
-
-    int ret = (int) mMount.read(mFileHandle, b, len, -1);
-    if (ret < 0) {
-      // attempt to reset to old position.
-      mMount.lseek(mFileHandle, mPos, CephMount.SEEK_SET);
-      throw new IOException("ceph.read: ret = " + ret);
-    }
-    mPos += ret;
-    return ret;
+    int totalRead = 0;
+    int initialLen = len;
+    int read;
+
+    do {
+      read = Math.min(len, mBufValid - mBufPos);
+      try {
+        System.arraycopy(mBuf, mBufPos, buf, off, read);
+      } catch (IndexOutOfBoundsException ie) {
+        throw new IOException(
+          "CephInputStream.read: Indices out of bounds:" + "read length is "
+          + len + ", buffer offset is " + off + ", and buffer size is "
+          + buf.length);
+      } catch (ArrayStoreException ae) {
+        throw new IOException(
+          "Uh-oh, CephInputStream.read failed to do an array"
```

Comment:
Hahaha, for production code, can you remove the "Uh-oh", although I like it hahaha